Package requirements: failure with compiler flags

### Steps to reproduce

With a `packages.yaml` like:

```
packages:
  zlib:
    require: cflags="-g"
```

`spack spec zlib` fails. Claiming that the requirements cannot be satisfied.

### Error message

```
...
spack.error.SpackError: concretization failed for the following reasons:

   1. Cannot satisfy the requirements in packages.yaml for package 'zlib'. You may want to delete them to proceed with concretization. To check where the requirements are defined run 'spack config blame packages'
```

### Information on your system

* `spack config blame packages` shows no non-default config (other than the one I explicitly add in this report)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

@scheibelp The issue with flags is that they're treated differently from variants, and there are currently no rules for `clingo` to generate a set of flags to satisfy the requirement. If you add something like:[^1]
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index e96e108e0e..57c8ff689c 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -461,6 +461,15 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
 % Config required semantics
 %-----------------------------------------------------------------------------
 
+{ attr("node_flag", A1, A2, A3) } :-
+  requirement_group_member(Y, Package, _),
+  condition_requirement(Y,"node_flag", A1, A2, A3).
+
+{ attr("node_flag_source", A1, A2, A3) } :-
+  requirement_group_member(Y, Package, _),
+  condition_requirement(Y,"node_flag_source", A1, A2, A3).
+
+
 package_in_dag(Package) :-  attr("node", Package).
 package_in_dag(Package) :-  attr("virtual_node", Package).

```
You'll get past the concretization error into an error in the builder:
```console
$ spack -d spec zlib
[ ... ]
==> [2023-04-25-16:10:52.079882] Failure reading URL: HTTP Error 404: Not Found
==> [2023-04-25-16:10:52.082865] Reading config from file /home/culpo/.spack/linux/compilers.yaml
==> [2023-04-25-16:10:52.187133] UNKNOWN SYMBOL: attr("node_compiler", zlib, gcc)
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack_installable/main.py", line 37, in main
    sys.exit(spack.main.main(argv))
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1004, in main
    return _main(argv)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 959, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 987, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 636, in _invoke_command
    return_val = command(parser, args)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/spec.py", line 103, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/__init__.py", line 222, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 2821, in concretize
    self._new_concretize(tests)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/spec.py", line 2796, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 2736, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 820, in solve
    answers = builder.build_specs(spec_attrs)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 2601, in build_specs
    self.reorder_flags()
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/asp.py", line 2519, in reorder_flags
    assert set(compiler_flags) == set(ordered_compiler_flags), msg
AssertionError: {'-g'} does not equal set()
```

[^1]: To be clear, I'm not sure these are the correct rules to be added, but they prove that part of the issue is the absence of any choice rule that let clingo add a flag to the answer set.

--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -2463,11 +2463,16 @@ class SpecBuilder(object):
 
                     # add flags from each source, lowest to highest precedence
                     for name in sorted_sources:
-                        source = self._specs[name] if name in self._hash_specs else cmd_specs[name]
-                        extend_flag_list(from_sources, source.compiler_flags.get(flag_type, []))
+                        all_src_flags = list()
+                        per_pkg_sources = [self._specs[name]]
+                        if name in cmd_specs:
+                            per_pkg_sources.append(cmd_specs[name])
+                        for source in per_pkg_sources:
+                            all_src_flags.extend(source.compiler_flags.get(flag_type, []))
+                        extend_flag_list(from_sources, all_src_flags)
 
                 # compiler flags from compilers config are lowest precedence
-                ordered_compiler_flags = from_compiler + from_sources
+                ordered_compiler_flags = list(llnl.util.lang.dedupe(from_compiler + from_sources))
                 compiler_flags = spec.compiler_flags.get(flag_type, [])
 
                 msg = "%s does not equal %s" % (set(compiler_flags), set(ordered_compiler_flags))

--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -2137,6 +2137,10 @@ class SpackSolverSetup(object):
 
         # get list of all possible dependencies
         self.possible_virtuals = set(x.name for x in specs if x.virtual)
+        for spec in specs:
+            if not spec.virtual:
+                for provided in spec.package_class.provided:
+                    self.possible_virtuals.add(provided.name)
         possible = spack.package_base.possible_dependencies(
             *specs, virtuals=self.possible_virtuals, deptype=spack.dependency.all_deptypes
         )
@@ -2463,11 +2467,16 @@ class SpecBuilder(object):
 
                     # add flags from each source, lowest to highest precedence
                     for name in sorted_sources:
-                        source = self._specs[name] if name in self._hash_specs else cmd_specs[name]
-                        extend_flag_list(from_sources, source.compiler_flags.get(flag_type, []))
+                        all_src_flags = list()
+                        per_pkg_sources = [self._specs[name]]
+                        if name in cmd_specs:
+                            per_pkg_sources.append(cmd_specs[name])
+                        for source in per_pkg_sources:
+                            all_src_flags.extend(source.compiler_flags.get(flag_type, []))
+                        extend_flag_list(from_sources, all_src_flags)
 
                 # compiler flags from compilers config are lowest precedence
-                ordered_compiler_flags = from_compiler + from_sources
+                ordered_compiler_flags = list(llnl.util.lang.dedupe(from_compiler + from_sources))
                 compiler_flags = spec.compiler_flags.get(flag_type, [])
 
                 msg = "%s does not equal %s" % (set(compiler_flags), set(ordered_compiler_flags))

--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -2137,10 +2137,6 @@ class SpackSolverSetup(object):
 
         # get list of all possible dependencies
         self.possible_virtuals = set(x.name for x in specs if x.virtual)
-        for spec in specs:
-            if not spec.virtual:
-                for provided in spec.package_class.provided:
-                    self.possible_virtuals.add(provided.name)
         possible = spack.package_base.possible_dependencies(
             *specs, virtuals=self.possible_virtuals, deptype=spack.dependency.all_deptypes
         )
@@ -2467,16 +2463,11 @@ class SpecBuilder(object):
 
                     # add flags from each source, lowest to highest precedence
                     for name in sorted_sources:
-                        all_src_flags = list()
-                        per_pkg_sources = [self._specs[name]]
-                        if name in cmd_specs:
-                            per_pkg_sources.append(cmd_specs[name])
-                        for source in per_pkg_sources:
-                            all_src_flags.extend(source.compiler_flags.get(flag_type, []))
-                        extend_flag_list(from_sources, all_src_flags)
+                        source = self._specs[name] if name in self._hash_specs else cmd_specs[name]
+                        extend_flag_list(from_sources, source.compiler_flags.get(flag_type, []))
 
                 # compiler flags from compilers config are lowest precedence
-                ordered_compiler_flags = list(llnl.util.lang.dedupe(from_compiler + from_sources))
+                ordered_compiler_flags = from_compiler + from_sources
                 compiler_flags = spec.compiler_flags.get(flag_type, [])
 
                 msg = "%s does not equal %s" % (set(compiler_flags), set(ordered_compiler_flags))

--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -501,11 +501,13 @@ requirement_group_satisfied(Package, X) :-
   requirement_group_member(Y, Package, _),
   activate_requirement_rules(Package),
   condition_requirement(Y,"node_flag", A1, A2, A3).
-
 { attr("node_flag_source", A1, A2, A3) } :-
   requirement_group_member(Y, Package, _),
   activate_requirement_rules(Package),
   condition_requirement(Y,"node_flag_source", A1, A2, A3).
+:- impose(ID), do_not_impose(ID).
+virtual(Virtual) :- possible_provider(Package, Virtual).
+attr("virtual_node", Virtual) :- provider(Package, Virtual).
 
 requirement_weight(Package, Group, W) :-
   W = #min {

--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -498,13 +498,13 @@ requirement_group_satisfied(Package, X) :-
 % flags if their only source is from a requirement. This is overly-specific
 % and should use a more-generic approach like in https://github.com/spack/spack/pull/37180
 { attr("node_flag", A1, A2, A3) } :-
-  requirement_group_member(Y, Package, _),
-  activate_requirement_rules(Package),
+  requirement_group_member(Y, Package, X),
+  activate_requirement(Package, X),
   condition_requirement(Y,"node_flag", A1, A2, A3).
 
 { attr("node_flag_source", A1, A2, A3) } :-
-  requirement_group_member(Y, Package, _),
-  activate_requirement_rules(Package),
+  requirement_group_member(Y, Package, X),
+  activate_requirement(Package, X),
   condition_requirement(Y,"node_flag_source", A1, A2, A3).
 
 requirement_weight(Package, Group, W) :-

--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -501,13 +501,11 @@ requirement_group_satisfied(Package, X) :-
   requirement_group_member(Y, Package, _),
   activate_requirement_rules(Package),
   condition_requirement(Y,"node_flag", A1, A2, A3).
+
 { attr("node_flag_source", A1, A2, A3) } :-
   requirement_group_member(Y, Package, _),
   activate_requirement_rules(Package),
   condition_requirement(Y,"node_flag_source", A1, A2, A3).
-:- impose(ID), do_not_impose(ID).
-virtual(Virtual) :- possible_provider(Package, Virtual).
-attr("virtual_node", Virtual) :- provider(Package, Virtual).
 
 requirement_weight(Package, Group, W) :-
   W = #min {

--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -500,12 +500,12 @@ requirement_group_satisfied(Package, X) :-
 { attr("node_flag", A1, A2, A3) } :-
   requirement_group_member(Y, Package, X),
   activate_requirement(Package, X),
-  condition_requirement(Y,"node_flag", A1, A2, A3).
+  imposed_constraint(Y,"node_flag_set", A1, A2, A3).
 
 { attr("node_flag_source", A1, A2, A3) } :-
   requirement_group_member(Y, Package, X),
   activate_requirement(Package, X),
-  condition_requirement(Y,"node_flag_source", A1, A2, A3).
+  imposed_constraint(Y,"node_flag_source", A1, A2, A3).
 
 requirement_weight(Package, Group, W) :-
   W = #min {

--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -494,6 +494,19 @@ requirement_group_satisfied(Package, X) :-
   activate_requirement(Package, X),
   requirement_group(Package, X).
 
+% TODO: the following two choice rules allow the solver to add compiler
+% flags if their only source is from a requirement. This is overly-specific
+% and should use a more-generic approach like in https://github.com/spack/spack/pull/37180
+{ attr("node_flag", A1, A2, A3) } :-
+  requirement_group_member(Y, Package, _),
+  activate_requirement_rules(Package),
+  condition_requirement(Y,"node_flag", A1, A2, A3).
+
+{ attr("node_flag_source", A1, A2, A3) } :-
+  requirement_group_member(Y, Package, _),
+  activate_requirement_rules(Package),
+  condition_requirement(Y,"node_flag_source", A1, A2, A3).
+
 requirement_weight(Package, Group, W) :-
   W = #min {
     Z : requirement_has_weight(Y, Z), condition_holds(Y), requirement_group_member(Y, Package, Group);

--- a/lib/spack/spack/test/concretize_requirements.py
+++ b/lib/spack/spack/test/concretize_requirements.py
@@ -381,6 +381,22 @@ packages:
     assert s2.satisfies("%gcc+shared")
 
 
+@pytest.mark.regression("34241")
+def test_require_cflags(concretize_scope, test_repo):
+    """Ensures that flags can be required from configuration."""
+    if spack.config.get("config:concretizer") == "original":
+        pytest.skip("Original concretizer does not support configuration" " requirements")
+
+    conf_str = """\
+packages:
+  y:
+    require: cflags="-g"
+"""
+    update_packages_config(conf_str)
+    spec = Spec("y").concretized()
+    assert spec.satisfies("cflags=-g")
+
+
 def test_requirements_for_package_that_is_not_needed(concretize_scope, test_repo):
     """Specify requirements for specs that are not concretized or
     a dependency of a concretized spec (in other words, none of

